### PR TITLE
Error messages

### PR DESCRIPTION
**Pull request by adam-fowler**
- Add messages to error responses returned by Hummingbird
- Return messages as json

### CODE REVIEW

File: Sources/Hummingbird/Files/FileMiddleware.swift [hunk 1]
```diff
@@ -98,12 +98,12 @@ public struct FileMiddleware<Context: BaseRequestContext, Provider: FileProvider
 
             // Remove percent encoding from URI path
             guard let path = request.uri.path.removingPercentEncoding else {
-                throw HTTPError(.badRequest)
+                throw HTTPError(.badRequest, message: "Invalid URL")
```
Comment (Joannis) [inline Sources/Hummingbird/Files/FileMiddleware.swift:101]:
```suggestion
                throw HTTPError(.badRequest, message: "Invalid percent encoding in URL")
```

File: Sources/Hummingbird/Files/FileMiddleware.swift [hunk 2]
```diff
@@ -98,12 +98,12 @@ public struct FileMiddleware<Context: BaseRequestContext, Provider: FileProvider
...
             }
 
             // file paths that contain ".." are considered illegal
             guard !path.contains("..") else {
-                throw HTTPError(.badRequest)
+                throw HTTPError(.badRequest, message: "URLs including '..' are not supported")
```
Comment (Joannis) [inline Sources/Hummingbird/Files/FileMiddleware.swift:106]: I don't think we should provide a message here, this is basically just used maliciously. Any info we give can only help attackers.

File: Sources/Hummingbird/Router/Parameters+UUID.swift [hunk 3]
```diff
@@ -33,7 +33,7 @@ extension Parameters {
         guard let param = self[s[...]],
               let result = UUID(uuidString: String(param))
         else {
-            throw HTTPError(.badRequest)
+            throw HTTPError(.badRequest, message: "Parameter '\(s)' can not be converted to the expected type (UUID)")
```
Comment (Joannis) [inline Sources/Hummingbird/Router/Parameters+UUID.swift:36]: We shouldn't expose the parameter name here, that's internal information. We can however expose the value provided. I'm also not sure we should provide the expected type per-say.

File: Sources/Hummingbird/Router/Parameters+UUID.swift [hunk 4]
```diff
@@ -53,7 +53,7 @@ extension Parameters {
     public func requireAll(_ s: String, as: UUID.Type) throws -> [UUID] {
         return try self[values: s[...]].map {
             guard let result = UUID(uuidString: String($0)) else {
-                throw HTTPError(.badRequest)
+                throw HTTPError(.badRequest, message: "One of the parameters '\(s)' can not be converted to the expected type (UUID)")
```
Comment (Joannis) [inline Sources/Hummingbird/Router/Parameters+UUID.swift:56]: Likewise here

File: Sources/Hummingbird/Router/Parameters.swift [hunk 5]
```diff
@@ -46,7 +46,7 @@ public extension Parameters {
     /// - Parameter s: parameter id
     func require(_ s: String) throws -> String {
         guard let param = self[s[...]].map({ String($0) }) else {
-            throw HTTPError(.badRequest)
+            throw HTTPError(.badRequest, message: "Expected parameter '\(s)' does not exist")
```
Comment (Joannis) [inline Sources/Hummingbird/Router/Parameters.swift:49]: Likewise here, we should consider not exposing too much info

**Review by Joannis — CHANGES_REQUESTED** [review]
I've got some doubts here, and want to be extra safe. Let's make sure we consider these more thoroughly.

File: Sources/Hummingbird/Files/FileMiddleware.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/Hummingbird/Files/FileMiddleware.swift]: removed

File: Sources/Hummingbird/Router/Parameters+UUID.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/Hummingbird/Router/Parameters+UUID.swift]: Removed parameter name and replaced with text

File: Sources/Hummingbird/Router/Parameters.swift (hunk 5 above)
Comment (adam-fowler) [reply Sources/Hummingbird/Router/Parameters.swift]: Removed parameter name and replaced with text

**Comment by adam-fowler** [comment]
> Also considering #432, we should consider making empty error messages (possibly optionally) non-json. I don't think it makes sense to serialize `{}` into an `application/json`

Errors with no body no longer report as having a JSON content type

**Review by Joannis — APPROVED** [review]

**Pull request merged**